lib/ansible/playbook/conditional.py

```diff
@@ -212,17 +213,32 @@ def generic_visit(self, node, inside_call=False, inside_yield=False):
             # to be looking for an undefined variable, return True, otherwise fail
             try:
                 # first we extract the variable name from the error message
-                var_name = re.compile(r"'(hostvars\[.+\]|[\w_]+)' is undefined").search(str(e)).groups()[0]
+                search_var = re.compile(r"'(hostvars\[.+\]|[\w_]+)' is undefined").search(to_text(e))
+                if not search_var and (re.compile(r"ansible.parsing.yaml.objects.AnsibleUnicode object has no element Undefined").search(to_text(e)) or
+                                       "list object has no element" in to_text(e)):
+                        var_name = None
+                else:
+                    var_name = search_var.groups()[0]
```

Comment:
There's one cornercase here.  The conditional that controls this can get here with ```search_var == None``` if the specific messages are also not present.  For instance:
``` python
e = 'This is a test'
search_var =  re.compile(r"'(hostvars\[.+\]|[\w_]+)' is undefined").search(to_text(e))
if not search_var and (re.compile(r"ansible.parsing.yaml.objects.AnsibleUnicode object has no element Undefined").search(to_text(e)) or "list object has no element" in to_text(e)):
    var_name = None
else:
    var_name = search_var.groups()[0]
```